pkg/contract-manager/reservation_controller.go

```diff
@@ -47,13 +55,309 @@ type ReservationReconciler struct {
+		return nil, err
+	}
+
+	// Send the POST request to the server
+	url := discoverymanager.SERVER_ADDR + "/reserveflavour/" + flavourID
+	klog.Infof("Sending POST request to %s", url)
+	resp, err := http.Post(url, "application/json", bytes.NewBuffer(selectorBytes))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	// Check if the response status code is 200 (OK)
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("received non-OK response status code: %d", resp.StatusCode)
+	}
+
+	// Read the response body
+	respBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var transaction Transaction
+	if err := json.Unmarshal(respBody, &transaction); err != nil {
+		return nil, err
+	}
+
+	return &transaction, nil
+}
+
+// purchaseFlavour purchases a flavour with the given flavourID
+func purchaseFlavour(ctx context.Context, transaction *Transaction, reservation *reservationv1alpha1.Reservation) (*ResponsePurchase, error) {
+	flavourID := transaction.FlavourID
+	body := map[string]interface{}{
+		"transactionID": transaction.TransactionID,
+		"flavourID":     flavourID,
+		"buyerID":       reservation.Spec.Buyer.NodeID,
+	}
+
+	selectorBytes, err := json.Marshal(body)
+	if err != nil {
+		return nil, err
+	}
+
+	// Send the POST request to the server
+	resp, err := http.Post(discoverymanager.SERVER_ADDR+"/purchaseflavour/"+flavourID, "application/json", bytes.NewBuffer(selectorBytes))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	// Check if the response status code is 200 (OK)
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("received non-OK response status code: %d", resp.StatusCode)
+	}
+
+	// Read the response body
+	respBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var purchase ResponsePurchase
+	if err := json.Unmarshal(respBody, &purchase); err != nil {
+		return nil, err
+	}
+
+	return &purchase, nil
+}
+
+// CreateContractCustomResource creates a Contract CR
+func createContractCustomResource(ctx context.Context, reservation *reservationv1alpha1.Reservation) *reservationv1alpha1.Contract {
+	contract := &reservationv1alpha1.Contract{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "contract-" + reservation.Name,
+			Namespace: discoverymanager.DEFAULT_NAMESPACE,
+		},
+		Spec: reservationv1alpha1.ContractSpec{
+			FlavourID: reservation.Spec.FlavourID,
+			Buyer: nodecorev1alpha1.NodeIdentity{
+				NodeID: reservation.Spec.Buyer.NodeID,
+			},
+			Seller: nodecorev1alpha1.NodeIdentity{
+				NodeID: reservation.Spec.Seller.NodeID,
+			},
+		},
+		Status: reservationv1alpha1.ContractStatus{
+			Phase: nodecorev1alpha1.PhaseStatus{
+				Phase: reservation.Status.Phase.Phase,
+			},
+		},
+	}
+
+	return contract
+}
+
+// setReservationPhase sets the phase of the discovery
+func setReservationPhase(r *reservationv1alpha1.Reservation, phase nodecorev1alpha1.Phase) {
+	r.Status.Phase.Phase = phase
+}
```

Comment:
Where is the `lastUpdateTime`, possibly the `message`, and the `startTime` update?

Reply:
Fixed, fixed, fixed!